# Module 3

Importing libraries

In [1]:
import pandas as pd #Data-Wrangling library
import numpy as np #Numerical Python library
import matplotlib.pyplot as plt #Data-Visualization library
import seaborn as sns #Data-Visualization library

## 3.1 

In [2]:
#Loading gold.csv
df_gold = pd.read_csv('Datasets/Gold/GOLD.csv')

In [3]:
#First five rows
df_gold.head()

Date  Price   Open   High    Low    Vol. Change %   Pred  \
0  May 04, 2017  28060  28400  28482  28025   0.08K   -1.79%  738.0   
1  May 05, 2017  28184  28136  28382  28135   0.06K    0.44% -146.0   
2  May 08, 2017  28119  28145  28255  28097   7.85K   -0.23%   30.0   
3  May 09, 2017  27981  28125  28192  27947  10.10K   -0.49%  357.0   
4  May 10, 2017  28007  28060  28146  27981   9.28K    0.09%  124.0   

          new  
0  117.570740  
1  295.430176  
2  132.123714  
3  101.298064  
4  112.153318

In [4]:
#Last five rows
df_gold.tail()

Date  Price   Open   High    Low   Vol. Change %  Pred  \
507  Apr 26, 2019  31868  31851  31934  31705  9.67K    0.08%   NaN   
508  Apr 30, 2019  31625  31800  31824  31597  6.44K   -0.76%   NaN   
509  May 01, 2019  31563  31604  31657  31503  1.55K   -0.20%   NaN   
510  May 02, 2019  31203  31420  31425  31160  0.48K   -1.14%   NaN   
511  May 03, 2019  31341  31250  31500  31163  0.08K    0.44%   NaN   

            new  
507  247.177322  
508   52.201158  
509  113.293305  
510   48.365693  
511  429.924911

In [5]:
#Statistical Summary
df_gold.describe()

Price          Open          High           Low         Pred  \
count    512.000000    512.000000    512.000000    512.000000   411.000000   
mean   30364.583984  30368.412109  30491.089844  30243.320312   341.712895   
std     1284.378623   1296.856656   1307.031684   1271.936410   271.075609   
min    27812.000000  27805.000000  27887.000000  27620.000000  -146.000000   
25%    29432.500000  29436.750000  29499.750000  29345.750000   156.500000   
50%    30427.500000  30399.500000  30545.000000  30286.000000   299.000000   
75%    31271.750000  31300.000000  31447.500000  31160.750000   475.000000   
max    33753.000000  34247.000000  34400.000000  33680.000000  2210.000000   

               new  
count   512.000000  
mean    245.161859  
std     205.788897  
min      11.109814  
25%     101.369172  
50%     184.537244  
75%     320.835031  
max    1407.321485

In [6]:
#Dimesions
df_gold.shape

(512, 9)

There are 512 observations and 9 variables.

In [7]:
#Datatype of each columns
df_gold.dtypes

Date         object
Price         int64
Open          int64
High          int64
Low           int64
Vol.         object
Change %     object
Pred        float64
new         float64
dtype: object

In [8]:
#Missing values in each column
df_gold.isnull().sum()

Date          0
Price         0
Open          0
High          0
Low           0
Vol.          0
Change %      0
Pred        101
new           0
dtype: int64

There are 101 missing values in Pred column

In [9]:
#Importing LinearRegression
from sklearn.linear_model import LinearRegression

In [10]:
#Creating linear regression object
lm = LinearRegression()
lm

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [11]:
#Storing x and y for predicting Pred column
X = df_gold[df_gold['Pred'].notnull()][['Price', 'Open', 'High', 'Low']]
Y = df_gold[df_gold['Pred'].notnull()]['Pred']

In [12]:
#Now we randomly select some part for training and some for testing

#import train_test_split
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size = 0.15, random_state = 1)

print("Size of trianing set: ", x_train.shape[0])
print("Size of testing set: ", x_test.shape[0])

Size of trianing set:  349
Size of testing set:  62


In [13]:
#Fitting the linear model
lm.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [14]:
#R^2 on train data
lm.score(x_train, y_train)

1.0

In [15]:
#R^2 on test data
lm.score(x_test, y_test)

1.0

In [16]:
#Intercept of linear model
lm.intercept_

7.747757990728132e-11

In [17]:
#Coefficients of linear model
lm.coef_

array([ 2.,  3., -1., -4.])

Our linear regression model is of the form-
                <b>Y_hat = 7.74e-11 + 2 * C + 3 * O -  1 * H  -  4 * L </b>

In [18]:
Pred_x = df_gold[df_gold['Pred'].isnull()][['Price', 'Open', 'High', 'Low']]
Pred_y = df_gold[df_gold['Pred'].isnull()]['Pred']

In [19]:
#Filling up the unfilled data
df_gold['Pred'] = lm.predict(df_gold[['Price', 'Open', 'High', 'Low']])

<b><i> The pred column is linearly related to Price, Open, High, and Low column of GOLD dataset

In [20]:
#Storing X and Y for the new column
X = df_gold[['Price', 'Open', 'High', 'Low']]
Y = df_gold[['new']]

In [21]:
#randomly splitting data into train and test
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size = 0.3, random_state = 3)

In [22]:
#Try to fit Linear model first
lm.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [23]:
#R^2 score on train data
lm.score(x_train, y_train)

0.9999870514485087

In [24]:
#R^2 score on test data
lm.score(x_test, y_test)

0.9999900146732249

Accuracy of our model is quite good. 99 percent of the variability in the outcome is being explained by our model.

In [25]:
#Fitting polynomial regression model

from sklearn.preprocessing import PolynomialFeatures
pr = PolynomialFeatures(degree = 3)

In [26]:
x_train_pr = pr.fit_transform(x_train)

In [27]:
poly = LinearRegression()

In [28]:
poly.fit(x_train_pr, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [29]:
#R^2 on train data
poly.score(x_train_pr, y_train)

0.9999999924829289

In [30]:
#R^2 on test data
x_test_pr = pr.fit_transform(x_test)
poly.score(x_test_pr, y_test)

0.9999999941401835

The accuracy of our model has been improved using the polynomial regression model of degree 3.

<b><i>new column of our dataset is polynomially related to Price, Open, High and Low columns of our dataset.

## 3.2 

In [31]:
#Loading Infosys stock

df_infy = pd.read_csv('Saved Datasets/week2.csv')

#Loading Nifty50 market index

df_nifty = pd.read_csv('Datasets/Nifty50/Nifty50.csv')

In [32]:
#Data Cleaning

#Changing datatype of Date column of both dataframes
df_infy['Date'] = df_infy['Date'].astype('datetime64')
df_nifty['Date'] = pd.to_datetime(df_nifty['Date'], dayfirst = True)

In [33]:
#Setting up Date as index of both dataframes

df_infy.set_index('Date', inplace = True)
df_nifty.set_index('Date', inplace = True)

In [34]:
#Merging up Close Price columns of both dataframes
df_beta = pd.merge(df_infy[['Close Price', 'Month', 'Year']], df_nifty['Close'], how = 'inner', left_index = True,
                   right_index = True)

In [35]:
#Renaming columns
df_beta.columns = ['Infy','Month', 'Year', 'Nifty50']

In [36]:
#Importing libraries to perform regression using OLS
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [37]:
#Building model for calculating daily Beta for last 3 months
model = smf.ols('Infy ~ Nifty50', data = df_beta.iloc[-90:-1]).fit()

In [38]:
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                   Infy   R-squared:                       0.066
Model:                            OLS   Adj. R-squared:                  0.056
Method:                 Least Squares   F-statistic:                     6.197
Date:                Mon, 08 Jun 2020   Prob (F-statistic):             0.0147
Time:                        11:25:41   Log-Likelihood:                -411.37
No. Observations:                  89   AIC:                             826.7
Df Residuals:                      87   BIC:                             831.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    535.5705     77.625      6.899      0.0

<b>Beta coefficient for last 3 months <br/></b>
The null hypothesis is that Beta = 0 for the last 3 months.<br/>
The Beta value calculated from our model is 0.0173 and the p-value associated with the t-stat of 2.489 is 0.015 which is lower than 0.05. So we reject the null hypothesis and there is relationship exists between the returns of market index Nifty50 and Infosys.<br/>
The value of R^2 statistic is 0.066 that means 6.6% of the variability in returns of Infosys is explained by the returns of Nifty. That indicates a very week relationship between the two of them.<br/>
The value of Beta is 0.0173 that means the increase of 10000 rupee in returns of Nifty will result in gain of 173 in the returns of Infosys. It also indicates that for the last 3 months shares of Infosys are less volatile relative to Nifty.

In [39]:
#Building model for monthly Beta value

#Grouping dataset on month
df_mbeta = df_beta.groupby(['Year', 'Month'])[['Infy', 'Nifty50']].mean()

#Model
model_2 = smf.ols('Infy ~ Nifty50', data = df_mbeta).fit()

#Summary
print(model_2.summary())

                            OLS Regression Results                            
Dep. Variable:                   Infy   R-squared:                       0.009
Model:                            OLS   Adj. R-squared:                 -0.034
Method:                 Least Squares   F-statistic:                    0.2204
Date:                Mon, 08 Jun 2020   Prob (F-statistic):              0.643
Time:                        11:25:45   Log-Likelihood:                -169.66
No. Observations:                  25   AIC:                             343.3
Df Residuals:                      23   BIC:                             345.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1339.4002    826.944      1.620      0.1

<b>Monthly Beta coefficient <br/></b>
The null hypothesis is that Beta = 0.<br/>
The Beta value calculated from our model is -0.0366 and the p-value associated with the t-stat of -0.469 is 0.643 which is greater than 0.05. So we accept the null hypothesis and there is no relationship exists between the returns of market index Nifty50 and Infosys.<br/>
The value of R^2 statistic is 0.009 that means 0.9% of the variability in returns of Infosys is explained by the returns of Nifty. That indicates a very week relationship between the two of them.<br/>
The value of Beta is -0.0366 that means the increase of 10000 rupee in returns of Nifty will result in loss of 366 in the returns of Infosys. It also indicates that the shares of Infosys are less volatile relative to Nifty.